### ISSUE TITLE
tree.diff_to_tree results in many unnecessary sys calls

### ISSUE THREAD

**Issue by Nukesor** [libgit2/pygit2#764]
Hi!

I'm currently working on a git scanner with a very high throughput for my bachelor thesis.
As the scanner runs with 8 threads, I'm observing severe slowdowns on my server.
After running the command with `strace` i noticed that there are a whole lot of unnecessary syscalls issued on each diff. It looks like  for each file in the diff, there is one `stat` for the actual file, two `access` for `/etc/gitattributes` and two `stat` for `/info/attributes`. 

This results in a lot of syscalls and slows things down significantly as the program seems to wait for those calls most of the time. The repositories are located inside `/tmp` and cleaned after each scan. There are no signs of RAM shortage as there are about 8GB+ unused space in the `/tmp` fs.

The repositories are cloned `bare`. The actual function call is:
```
diff = commit.tree.diff_to_tree(commit.parents[0].tree)
self.commit_stats[commit.hex] = {
    'additions': diff.stats.insertions,
    'deletions': diff.stats.deletions,
}
```

I'm not sure if this is the right place or if this is rather a libgit2 issue.
Anyway I'm happy for any help as I'm not that familiar with the pythin-C binding mechanics.

Relevant strace output example:
```
[pid 32220] stat("src/main.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/main.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/states/Level.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/states/Level.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/main.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/main.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/states/Level.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("src/states/Level.lua", 0x7ffd6ef8c410) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] stat("/tmp/gitalizer/C-gotoh/atomazed/info/attributes", 0x7ffd6ef8c250) = -1 ENOENT (No such file or directory)
[pid 32220] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
```

**Comment by ethomson** [comment]
Indeed, this is an issue in libgit2 itself.  It should have recently been fixed by https://github.com/libgit2/libgit2/pull/4462.

**Comment by Nukesor** [comment]
Thanks for the quick response, this saved me a few hours of digging into libgit2!! Should i leave this issue open until the new libgit2 release is published? Otherwise feel free to close it..

**Comment by ethomson** [comment]
Sorry, I just wandered in from libgit2-land to drop this line.  😀   I'll let a pygit2 maintainer comment directly on the inclusion of a newer libgit2.  (Or if this change should have landed in pygit2 already - it's possible that we missed a code path that needs the same optimization.)

**Comment by Nukesor** [comment]
I just tried it out with the `master` libgit2 build and it looks a lot better.
There are still some `stat` calls for files, which shouldn't be necessary on a bare repository, but there aren't any syscalls for `.gitattributes` or `attributes` any longer.

Current output:
```
[pid 15375] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 15375] stat("/tmp/gitalizer/phoemur/slackbuilds-1/info/attributes", 0x7ffe5758e5f0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/README", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/README", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/README", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
```

**Comment by jdavid** [comment]
@ethomson When there will be a libgit2 release including this fix?

@Nukesor Those remaining calls to ``stat``, do you know whether they come from libgit2 or pygit2?

**Comment by Nukesor** [comment]
@jdavid As far as I can see they are issued from libgit2.

**Comment by Nukesor** [comment]
The issue is fixed on master libgit2

**Comment by andschwa** [comment]
Not to dig up an old issue, but did this perhaps regress? I am using the development version of pygit2 at eb4984e33f5334621358de5969446546b32a63cb with libgit2 v0.99.0 and am getting `diff.deltas` on a bare repo, but when I attached strace I saw tons of stat calls to the repos `info/attributes` file (returning `ENOENT`) and to all my `~/.gitconfig` and `~/.config/git/config`.

If this is expected behavior, is it possible to cache these lookups once per "session"?
